animation_nodes/nodes/sound/sound_spectrum.py

```diff
@@ -0,0 +1,152 @@
+
+    def create(self):
+        self.newInput("Sound", "Sound", "sound")
+        self.newInput("Float", "Frame", "frame")
+        self.newInput("Float", "Attack", "attack", value = 0.005, minValue = 0, maxValue = 1)
+        self.newInput("Float", "Release", "release", value = 0.6, minValue = 0, maxValue = 1)
+        self.newInput("Float", "Amplitude", "amplitude", value = 10)
+
+        if self.samplingMethod == "EXP":
+            self.newInput("Integer", "Count", "count", value = 20, minValue = 1)
+            self.newInput("Float", "Low", "low", value = 0, minValue = 0, maxValue = 1, hide = True)
+            self.newInput("Float", "High", "high", value = 1, minValue = 0, maxValue = 1, hide = True)
+            self.newInput("Float", "K", "k", value = 5, minValue = 0.00001)
+        elif self.samplingMethod == "CUSTOM":
+            self.newInput("Float List", "Pins", "pins")
+        elif self.samplingMethod == "SINGLE":
+            self.newInput("Float", "Low", "low", value = 0, minValue = 0, maxValue = 1)
+            self.newInput("Float", "High", "high", value = 0.1, minValue = 0, maxValue = 1)
+
+        self.newInput("Scene", "Scene", "scene", hide = True)
+
+        if self.samplingMethod in ("EXP", "CUSTOM", "FULL"):
+            self.newOutput("Float List", "Spectrum", "spectrum")
+        else:
+            self.newOutput("Float", "Spectrum", "spectrum")
+
+    def draw(self, layout):
+        layout.prop(self, "samplingMethod", text = "")
+
+    def drawAdvanced(self, layout):
+        layout.prop(self, "reductionFunction", text = "")
+        layout.prop(self, "smoothingSamples")
+        layout.prop(self, "kaiserBeta")
+        layout.prop(self, "minDuration")
+
+    def getExecutionFunctionName(self):
+        if self.samplingMethod == "EXP": return "executeExponential"
+        elif self.samplingMethod == "CUSTOM": return "executeCustom"
+        elif self.samplingMethod == "SINGLE": return "executeSingle"
+        elif self.samplingMethod == "FULL": return "executeFull"
+
+    def executeExponential(self, sound, frame, attack, release, amplitude, count, low, high, k, scene):
+        if len(sound.soundSequences) == 0: self.raiseErrorMessage("Empty sound!")
+        if not isValidRange(low, high): self.raiseErrorMessage("Invalid interval!")
+        if count < 1: self.raiseErrorMessage("Invalid count!")
+
+        spectrum = sound.computeTimeSmoothedSpectrum(frame / scene.render.fps,
+            frame / scene.render.fps + max(1 / scene.render.fps, self.minDuration),
+            attack, release, self.smoothingSamples, self.kaiserBeta)
+        maxFrequency = len(spectrum) - 1
+
+        scale = expm1(k) / (high - low)
+        pins = [low + expm1(k * i / count) / scale for i in range(count + 1)]
+
+        bars = DoubleList(count)
+        reductionFunction = reductionFunctions[self.reductionFunction]
+        for i in range(count):
+            x, y = int(pins[i] * maxFrequency), int(pins[i + 1] * maxFrequency)
```

Comment:
I think the computation of the frequency bins could be moved to a separate function. This function should be a generator function that yields the start and end of every frequency bin.

I believe that this way, the `executeExponential`, `executeCustom` and `executeFull` methods could share more code. Currently there is quite a lot of code duplication going on if I see that correctly.

Reply:
Alright. Time to know how to use `yield`.

Reply:
hmmm, looking at this again, each execution function have different inputs, so we can't join them into a single function. Do you mean that we should use a `getExecutionCode` function? Moreover, there are slight differences in the for loop used to compute the frequency bins, so we can use a unified generator.
Can you elaborate?

Reply:
Ok, but three cases? The `executeFull` is very different from the other two. I think including it in the helper function would needlessly complicate it, don't you think? After all, it is just two lines of code.

Reply:
hmm ok, leave it as is.